core/src/main/java/org/apache/brooklyn/core/typereg/TypePlanTransformers.java

```diff
@@ -0,0 +1,160 @@
+import com.google.common.collect.ArrayListMultimap;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Iterables;
+import com.google.common.collect.Multimap;
+
+public class TypePlanTransformers {
+
+    private static final Logger log = LoggerFactory.getLogger(TypePlanTransformers.class);
+
+    private static Collection<BrooklynTypePlanTransformer> getAll() {
+        return ImmutableList.copyOf(ServiceLoader.load(BrooklynTypePlanTransformer.class));
+    }
+
+    private static Collection<Class<? extends BrooklynTypePlanTransformer>> OVERRIDE;
+    @SafeVarargs
+    @VisibleForTesting
+    public synchronized static void forceAvailable(Class<? extends BrooklynTypePlanTransformer> ...classes) {
+        OVERRIDE = Arrays.asList(classes);
+    }
+    public synchronized static void clearForced() {
+        OVERRIDE = null;
+    }
+
+    public static Collection<BrooklynTypePlanTransformer> all(ManagementContext mgmt) {
+        // TODO cache these in the TypeRegistry, looking for new ones periodically or supplying a way to register them
+        Collection<Class<? extends BrooklynTypePlanTransformer>> override = OVERRIDE;
+        Collection<BrooklynTypePlanTransformer> result = new ArrayList<BrooklynTypePlanTransformer>();
+        if (override!=null) {
+            for (Class<? extends BrooklynTypePlanTransformer> o1: override) {
+                try {
+                    result.add(o1.newInstance());
+                } catch (Exception e) {
+                    Exceptions.propagate(e);
+                }
+            }
+        } else {
+            result.addAll(getAll());
+        }
+        for(BrooklynTypePlanTransformer t : result) {
+            t.injectManagementContext(mgmt);
+        }
+        return result;
+    }
+
+    /** returns a list of {@link BrooklynTypePlanTransformer} instances for this {@link ManagementContext}
+     * which may be able to handle the given plan; the list is sorted with highest-score transformer first */
+    @Beta
+    public static List<BrooklynTypePlanTransformer> forType(ManagementContext mgmt, RegisteredType type, RegisteredTypeLoadingContext constraint) {
+        Multimap<Double,BrooklynTypePlanTransformer> byScoreMulti = ArrayListMultimap.create(); 
+        Collection<BrooklynTypePlanTransformer> transformers = all(mgmt);
+        for (BrooklynTypePlanTransformer transformer : transformers) {
+            double score = transformer.scoreForType(type, constraint);
+            if (score>0) byScoreMulti.put(score, transformer);
+        }
+        Map<Double, Collection<BrooklynTypePlanTransformer>> tree = new TreeMap<Double, Collection<BrooklynTypePlanTransformer>>(byScoreMulti.asMap());
+        List<Collection<BrooklynTypePlanTransformer>> highestFirst = new ArrayList<Collection<BrooklynTypePlanTransformer>>(tree.values());
+        Collections.reverse(highestFirst);
+        return MutableList.copyOf(Iterables.concat(highestFirst)).asUnmodifiable();
```

Comment:
Could use `ImmutableList.copyOf()` instead.
Don't understand the usage of `concat` here.